### ISSUE TITLE
Administrator model customization

### ISSUE THREAD

**Issue by mdibaiee** [strapi/strapi#3399]
**Describe the bug**
Currently it seems it's not possible to create relations with the Administrator model (one-way).

An example use case:
We have a plugin in which we want to create new entities of a certain type, but we also want to know which administrator created them, so we would like to save a relationship between this entity and administrator. This can be a one-way relationship, but it seems even that is not possible.

**Steps to reproduce the behavior**
1. Create a model
2. Create a relationship with administrator:
```
"created_by": {
  "model": "administrator",
  "plugin": "admin"
}
```
3. Run project

**Expected behavior**
Ability to create relationships (even if one-way) to administrator

**Screenshots**
N/A

**Code snippets**
N/A

**System**
- Node.js version: v10.2.1
- NPM version: 5.6.0
- Strapi version: 3.0.0-beta.2
- Database: PostgreSQL
- Operating system: Archlinux

**Additional context**
N/A

**Comment by derrickmehaffy** [comment]
Agreed, @lauriejim this is also related to what I mentioned to you in slack about overriding the admin model file

**Comment by alexandrebodin** [comment]
Hi,

I agree with you it would be a nice feature to have. 

Right now you can implement your usecase by using the user from the users-permissions plugin (I know it means having duplicates of your admins in the users) but this will at least allow you to implement until this feature is implemented somehow.

@derrickmehaffy @mdibaiee What can of relations would you like to have ?

**Comment by cschweda** [comment]
I have a similar issue with the new way the admin is configured.

## My use case

We have an in-house, agency calendar. Users must sign in to view the calendar. The sign in authentication is handled by Strapi.

Until recently, the calendar user's roles were either Strapi 'Administrators' or Strapi 'Authenticated'. There is no 'Public' access to the calendar. Users registered via a registration link (the registration process handled by Strapi) -- and a successful, email verified registration meant you were automatically set as 'Authenticated'.

If your Strapi user role is 'authenticated', you're only able to view the calendar. If, however, your Strapi user role is 'Administrator', you're able to view the calendar as well as make edits to the calendar via the Strapi admin back-end. A unit manager would say, hey, so-and-so needs to edit the calendar as well as view the calendar. When I got this request, I manually bumped the user's role from 'Authenticated' to 'Administrator' -- and then I'd send that new 'Administrator' user a link to the Strapi admin.

The point here is that both 'Authenticated' and 'Administrator' users used the same login process -- go to the registration link on the front-end, register, verify e-mail, and then login via the ```/auth/local```. The best part of this is that the user -- no matter if she were 'Authenticated' or 'Administrator' could use the same username and password for access to the back-end Strapi admin and the front-end calendar app (written in Vue)

## What's happening now

Now, 'Administrator' roles have to be authenticated via ```/admin/auth/local``` on the front-end whereas 'Authenticated` users authenticate via ```/auth/local```.

## The issue

So the issue right now is that those admin users can't simply login to the calendar anymore. The ```/auth/local``` route won't authenticate them. They need a seperate username/password to access the Strapi backend. So they'll have two username/password combos -- one to access the Strapi back-end, another to access the 'front-end'.

## Confusion

So, I'm confused now. I'd been using the single route to login. Now there's two routes. So either I need to create sepeate usernames and passwords -- or I ask the user up front: are you an admin? Are you a user? ... and then set the auth route accordintly.

I understand the desire to seperate the strapi roles from the app roles -- but in my case, and (I suspect) many others -- the roles are identical. An 'admin' is an 'admin' -- so she should be able to log in to the strapi backend and login to the app front-end -- all with the same username and password.

**Comment by mdibaiee** [comment]
@alexandrebodin currently, our use case is just a one-way relationship to the administrator model.

We want our local plugins to be able to create models, which, as a metadata, include the Administrator which created them in their database record. So our table looks like this:

```
{
  "connection": "default",
  "collectionName": "sms",
  "info": {
    "name": "sms",
    "description": ""
  },
  "options": {
    "increments": true,
    "timestamps": true,
    "comment": ""
  },
  "attributes": {
    "created_by": {
      "model": "administrator",
      "plugin": "admin"
    },
    "body": {
      "type": "string",
      "required": true
    },
    "retailer": {
      "model": "retailer"
    },
    "users_count": {
      "type": "integer",
      "min": 0
    },
    "created_at": {
      "required": true,
      "type": "date"
    },
    "updated_at": {
      "required": true,
      "type": "date"
    },
    "deleted_at": {
      "required": false,
      "type": "date"
    }
  }
}
```

If I understand correctly, the only change I'm looking for right now is the ability to specify `admin` as the `plugin` (or any other type of change) to use the model.

If you can point me to where the code for this resides, I might be able to create a pull-request myself.

As a next step (and I can foresee this requirement), I think we potentially will want to customize the administrator model as well.

**Comment by alexandrebodin** [comment]
Hi guys. 

Here is how I think the admin / user-permissions should work in the future so we can discuss the short term wins we can have:

## User-permissions

The idea is to make the `user-permissions` plugin a real `oauth` / `open id connect` server and use passportjs to allow for a lot of providers. 

## Admin connection

In the same way I want the admin to use passport so people can configure it to connect to the users-permissions plugin if you want to. Use a ldap or any other passport providers.


What do you guys think about it ?

**Comment by alexandrebodin** [comment]
For the fields/columns customization it is a much simpler usecase. You can already custmize it by make the updates in the model directly in your bootstrap function.

**Comment by derrickmehaffy** [comment]
> For the fields/columns customization it is a much simpler usecase. You can already custmize it by make the updates in the model directly in your bootstrap function.

@alexandrebodin Probably should elaborate on this for over-riding the model.

**Comment by alexandrebodin** [comment]
You can add custom fields attributes (⚠️ not relations ⚠️ ) by creating a custom hook.

Make sure to put run it first by updating the config in 

`config/hook.json`

```json
{
  "timeout": 3000,
  "load": {
    "before": ["custom-hook"],
    "order": [
      "Define the hooks' load order by putting their names in this array in the right order"
    ],
    "after": []
  }
}
```
`hooks/custom-hook/index.js`
```js
const _ = require('lodash');

module.exports = () => {
  return {
    initialize(cb) {
      _.merge(strapi.admin.models.administrator, {
        attributes: {
          randomColum: {
            type: 'string',
          },
        },
      });

      cb();
    },
  };
};
```

We either can allow modifiy the model via json like the other and in the interface but this will make the possbile errors too important in my opinion.

I would like to provide and fairly simple api to customize a model on startup.

example:
```js
strapi.models('administrator').extend({
  atttributes: {
    newField: {},
    newRelation: {},
  },
});
```

**Comment by derrickmehaffy** [comment]
@alexandrebodin what about modifying existing values? Such as these: 

https://github.com/strapi/strapi/blob/15daa6212c1323c1ccb12099df92d8a51a704ee1/packages/strapi-admin/models/Administrator.settings.json#L2-L6

**Comment by derrickmehaffy** [comment]
And for example what if someone wants to modify their services/controllers of the admin package?

IE: https://github.com/strapi/strapi/blob/master/packages/strapi-admin/services/auth.js

I'm more for the opinion that the admin package needs to be split, one for the UI, the other should be an authentication plugin similar to `users-permissions`

**Comment by derrickmehaffy** [comment]
(or convert the whole admin package itself into a plugin)

**Comment by alexandrebodin** [comment]
you can change any meta with the example I provided. At your own risks.

I really don't think people should modify the controllers or services. Extending the model makes sense. Providing some programtic extension points too but just changing the files is really the best way to allo people to break everything at anytime. It doesn't really scale.
If it is missing features I would rather have them start a conversation and make a PR so it helps not only them but the community. Alllow people to just overwritte any files kinda break the idea of making Strapi open source as no one would need to contribute back to it 🤔 I guess it is a matter of opinion anyway.

**Comment by derrickmehaffy** [comment]
@alexandrebodin 

> Alllow people to just overwritte any files kinda break the idea of making Strapi open source as no one would need to contribute back to it

I would like to point out that realistically Strapi should for easy customization in all aspects, this should include options for all packages as not everyone's use-case is going to follow "the heard", and there is going to be some niche cases. As an example extending the admin authentication to add certain password, username, email restrictions. Maybe specific password polices, the options are limitless and hence the reason I think that admin package needs to allow for this. 

If we are allowing users to modify the "frontend" of the adminUI, the backend should have the same respect.

![image](https://user-images.githubusercontent.com/8593673/59344633-57bc9980-8cc3-11e9-93a4-4235657eb9ab.png)

**Comment by derrickmehaffy** [comment]
^^^ "The API" in the above image should be including the Admin Authentication API along with it's services, controllers, and routes.

**Comment by alexandrebodin** [comment]
As long as we can create apis that respect the Open/closed principle sure

**Comment by mdibaiee** [comment]
@alexandrebodin I would like to work on this since our product has a strict requirement for customizing the admin model. Can you please point me in a direction that matches your vision on this issue and I can get started on this?

**Comment by alexandrebodin** [comment]
Hi @mdibaiee before starting we need to figure out what is going to be allowed to be modified and how. 
Does having the Model in the UI is necessary for example. Or can we just give a way to extend it programatically.

**Comment by mdibaiee** [comment]
@alexandrebodin I agree. So, the way we think about it is that we want it to be a model just like any other model, very similar to how plugins are handled, we should be able to extend it and create relations with it. We would also like to manage administrators from the admin panel.

I think having it in the UI makes sense, if I try to think of a CMS, I can always imagine it having a panel for managing admin users (WordPress does that).

I also don't see a reason not to allow _extensions_ to the admin panel. Though one may argue that we should not be able to remove anything or modify the default attributes from the model to avoid breaking the application, and I think that makes sense.

I'm interested to hear your opinions on this as well.

**Comment by mdibaiee** [comment]
Any thoughts on this, @alexandrebodin ?

**Comment by justinluk** [comment]
I'd love a way to be able to associate an administrator with one of my content types.

**Comment by aefox** [comment]
> You can add custom fields attributes (⚠️ not relations ⚠️ ) by creating a custom hook.

@alexandrebodin I've just tried your suggestion and unfortunately the field does not get saved in mongo. The field correctly appears in the UI and the value I set for it is send as part of the PUT request to update an existing administrator (`/content-manager/explorer/strapi::administrator/:id`). But the value never reaches mongo. Any idea what could be wrong?

I'm on 3.0.0-beta.19.3

**Comment by quynguyencanalcircle** [comment]
> > You can add custom fields attributes (⚠️ not relations ⚠️ ) by creating a custom hook.
> 
> @alexandrebodin I've just tried your suggestion and unfortunately the field does not get saved in mongo. The field correctly appears in the UI and the value I set for it is send as part of the PUT request to update an existing administrator (`/content-manager/explorer/strapi::administrator/:id`). But the value never reaches mongo. Any idea what could be wrong?
> 
> I'm on 3.0.0-beta.19.3

I follow this way override the strapi-admin package for adding the relation field 'role' to the administrator model.

 First, I added one field to this file:
https://github.com/strapi/strapi/blob/master/packages/strapi-admin/models/Administrator.settings.json
  ```
 "role": {
      "plugin": "users-permissions",
      "model": "role"
    }
```
Then check the create and update functions on https://github.com/strapi/strapi/blob/master/packages/strapi-admin/controllers/Admin.js

```
const { email, username, password, blocked, ...otherProps } = ctx.request.body;
...
const user = {
      email: email,
      username: username,
      blocked: blocked === true ? true : false,
      password: await strapi.admin.services.auth.hashPassword(password),
      ...otherProps
    };
```
And it worked!
Hope it helps,

**Comment by leonelllagumbay** [comment]
Can i use the following approach. Im new to strapi

// helper/admin-overrides.js
const fs = require('fs');
 let dir = './extensions/admin/models/Administrator.settings.json';
 let dirDist = './node_modules/strapi-admin/models/Administrator.settings.json'
 fs.copyFileSync(dir, dirDist);

 dir = './extensions/admin/controllersx/Admin.js';
 dirDist = './node_modules/strapi-admin/controllers/Admin.js'
 fs.copyFileSync(dir, dirDist);

And use this in package.json
"scripts": {
    "develop": "npm run override-admin && strapi develop",
    "start": "npm run override-admin && strapi start",
    "override-admin": "node helper/admin-overrides.js",

**Comment by lauriejim** [comment]
This issue has been mentioned on **Strapi Community**. There might be relevant details there:

https://forum.strapi.io/t/extend-user-admin-model/713/1

**Comment by derrickmehaffy** [comment]
Due to the nature of how the admin plugin now works in v4, we currently do not plan to allow for this type of customization by default and there are other methods such as patch-package that do make it possible without needing support from us on this.

Marking as closed as discussed with @Aurelsicoko
